ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ObjectDiffusion/Inbound.hs

```diff
@@ -0,0 +1,469 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -Wno-partial-fields #-}
+
+module Ouroboros.Consensus.MiniProtocol.ObjectDiffusion.Inbound
+  ( objectDiffusionInbound
+  , TraceObjectDiffusionInbound (..)
+  , ObjectDiffusionInboundError (..)
+  , ProcessedObjectCount (..)
+  ) where
+
+import Cardano.Prelude (catMaybes)
+import Control.Concurrent.Class.MonadSTM.Strict.TVar.Checked
+import Control.Exception (assert)
+import Control.Monad (when)
+import Control.Monad.Class.MonadSTM
+import Control.Monad.Class.MonadThrow
+import Control.Tracer (Tracer, traceWith)
+import Data.Foldable as Foldable (foldl', toList)
+import Data.List qualified as List
+import Data.List.NonEmpty qualified as NonEmpty
+import Data.Map.Strict (Map)
+import Data.Map.Strict qualified as Map
+import Data.Sequence.Strict (StrictSeq)
+import Data.Sequence.Strict qualified as Seq
+import Data.Set (Set)
+import Data.Set qualified as Set
+import GHC.Generics (Generic)
+import Network.TypedProtocol.Core (N (Z), Nat (..), natToInt)
+import NoThunks.Class (NoThunks (..), unsafeNoThunks)
+import Ouroboros.Consensus.MiniProtocol.ObjectDiffusion.ObjectPool.API
+import Ouroboros.Network.NodeToNode.Version (NodeToNodeVersion)
+import Ouroboros.Network.Protocol.ObjectDiffusion.Inbound
+import Ouroboros.Network.Protocol.ObjectDiffusion.Type
+
+-- Note: This module is inspired from TxSubmission inbound side.
+
+data ProcessedObjectCount objectId object = ProcessedObjectCount
+  { pocAcceptedObjectsCount :: Int
+  -- ^ Just accepted this many objects.
+  , pocRejectedObjectsCount :: Int
+  -- ^ Just rejected this many objects.
+  }
+  deriving (Eq, Show)
+
+-- TODO: should we improve the tracing in adding more detail when e.g. IDs are
+-- received ?
```

Comment:
I think it is fine to leave it be an on-demand kind of things. Namely when someone finds a bug and feels the need to add tracing, they do so and then actually bundle this into a PR improving the tracing. I don't think it makes much sense to make sure we trace many things that we don't know people will use.